NAMA : YASMINE ALIFTA

NIM : 2311110044

KELAS : S1SD-04-01

# **AKTIVITAS PARTISIPATIF #1**

---



**MEMANGGIL DATASET & MENAMPILKAN DATA**

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Menampilkan semua kolom dataset
df = pd.read_excel('/content/CAD-Dataset.xlsx')
pd.set_option('display.max_columns', None)

# Menampilkan 20 baris pertama dataset
df.head(20)

No  Age  Weight  Length    Sex  BMI DM HTN Current Smoker EX-Smoker FH  \
0    1   53      90     175   Male    ?  n   Y              Y         n  n   
1    2   67      70     157  Fmale  NaN  n   Y              n         n  n   
2    3   54      54     164   Male  NaN  n   n              Y         n  n   
3    4   66      67     158  Fmale  NaN  n   Y              n         n  n   
4    5   50      87     153  Fmale  NaN  n   Y              n         n  n   
5    6   50      75     175   Male  NaN  n   n              Y         n  n   
6    7   55      80     165   Male  NaN  n   n              n         Y  n   
7    8   72      80     175   Male  NaN  Y   n              Y         n  n   
8    9   58      84     163  Fmale  NaN  n   n              n         n  n   
9   10   60      71     170   Male  NaN  Y   n              n         n  n   
10  11   58      75     168   Male  NaN  n   Y              n         Y  n   
11  12   80      67     153  Fmale  NaN  n   Y              n         n  n   
12  13   70      70     151  Fmale  NaN  Y   Y              n         n  n   
13  14   67      63     154  Fmale  NaN  Y   Y              n         n  n   
14  15   66      63     155  Fmale  NaN  Y   Y              n         n  n   
15  16   59      81     167   Male  NaN  Y   n              n         n  n   
16  17   41      68     169   Male  NaN  n   n              Y         n  n   
17  18   68      59     161  Fmale  NaN  n   n              n         n  Y   
18  19   60      89     163  Fmale  NaN  Y   Y              n         n  n   
19  20   65      72     150  Fmale  NaN  Y   Y              n         n  n   

   Obesity CRF CVA Airway disease Thyroid Disease CHF DLP   BP   PR Edema  \
0        Y   N   N              N               N   N   Y  110   80    NO   
1        Y   N   N              N               N   N   N  140   80   Yes   
2        N   N   N              N               N   N   N  100  100    NO   
3        Y   N   N              N               N   N   N  100   80    NO   
4        Y   N   N              N               N   N   N  110   80    NO   
5        N   N   N              N               N   N   N  118   70    NO   
6        Y   N   N              N               N   N   N  110   80    NO   
7        Y   N   N              N               N   N   Y  130   70    NO   
8        Y   N   N              N               N   N   N   90   50    NO   
9        N   N   N              N               N   N   N  130   70    NO   
10       Y   N   N              N               N   N   N  170   70    NO   
11       Y   N   N              N               ?   N   Y  140  100    NO   
12       Y   N   N              N               ?   N   Y  140   84    NO   
13       Y   N   N              N               N   N   Y  150   74    NO   
14       Y   N   N              N               N   N   Y  130   80    NO   
15       Y   N   N              N               N   N   Y  120   70    NO   
16       N   N   N              N               N   N   N  130   80    NO   
17       N   N   N              N               N   N   Y  115   70    NO   
18       Y   N   N              N               N   N   N  110   74    NO   
19       Y   Y   N              N               N   N   Y  150   70    NO   

   Weak Peripheral Pulse Lung rales Systolic Murmur Diastolic Murmur  \
0                      N          N               N                N   
1                      N          N               N                N   
2                      N          N               N                N   
3                      N          N               N                Y   
4                      N          N               Y                N   
5                      N          N               N                N   
6                      N          N               Y                N   
7                      N          N               N                N   
8                      N          N               N                N   
9                      N   

In [3]:
# menampilkan info dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 56 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   No                     303 non-null    int64  
 1   Age                    303 non-null    int64  
 2   Weight                 303 non-null    int64  
 3   Length                 303 non-null    int64  
 4   Sex                    303 non-null    object 
 5   BMI                    1 non-null      object 
 6   DM                     302 non-null    object 
 7   HTN                    303 non-null    object 
 8   Current Smoker         299 non-null    object 
 9   EX-Smoker              303 non-null    object 
 10  FH                     303 non-null    object 
 11  Obesity                303 non-null    object 
 12  CRF                    303 non-null    object 
 13  CVA                    303 non-null    object 
 14  Airway disease         303 non-null    object 
 15  Thyroi

Pada dataset tersebut, terdapat 303 baris dan 56 kolom.

**DATA PREPROCESSING**

Karena pada dataset terdapat beberapa tanda "?" sehinggga dataset tidak bisa diproses lebih lanjut. Agar bisa dilakukan preprocessing maka akan diubah menjadi NaN yang menjadi missing value, jadi dataset bisa di proses dengan mengisi missing value nya nanti.

In [4]:
# Mengganti tanda "?" dengan NaN
df.replace("?", np.nan, inplace=True)

<ipython-input-4-4829f373364e>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df.replace("?", np.nan, inplace=True)


In [5]:
# melihat adanya missing value
df.isnull().sum()

No                         0
Age                        0
Weight                     0
Length                     0
Sex                        0
BMI                      303
DM                         1
HTN                        0
Current Smoker             7
EX-Smoker                  0
FH                         0
Obesity                    0
CRF                        2
CVA                        5
Airway disease             0
Thyroid Disease            2
CHF                        0
DLP                        0
BP                         2
PR                         0
Edema                      0
Weak Peripheral Pulse      0
Lung rales                 0
Systolic Murmur            0
Diastolic Murmur           0
Typical Chest Pain         0
Dyspnea                    0
Function Class             0
Atypical                   0
Nonanginal                 4
Exertional CP              9
LowTH Ang                  4
Q Wave                     2
St Elevation               1
St Depression              2
Tinversion                 0
LVH                        0
Poor R Progression         0
FBS                        2
CR                         0
TG                         1
LDL                        2
HDL                        0
BUN                        1
ESR                        1
HB                         0
K                          0
Na                         0
WBC                        0
Lymph                      0
Neut                       0
PLT                        1
EF-TTE                     0
Region RWMA                0
VHD                        0
Cath                       0
dtype: int64

Dataset tersebut memiliki beberpa missing value, diantaranya:

- BMI: 303
- DM: 1
- Current Smoker: 7
- EX-Smoker: 0
- CRF: 2
- CVA: 5
- Thyroid Disease: 2
- DLP: 0
- BP: 2
- Nonanginal: 4
- Exertional CP: 9
- LowTH Ang: 4
- Q Wave: 2
- St Elevation: 1
- St Depression: 2
- FBS: 2
- TG: 1
- LDL: 2
- BUN: 1
- ESR: 1
- PLT: 1





Terlihat bahwa kolom yang memiliki paling banyak missing value yaitu kolom BMI sebanyak 303 yang merupakan jumlah keseluruhan baris dari dataset. BMI merupakan informasi yang penting karena bisa memberikan kontribusi penting dalam prediksi penyakit. Sehingga perlu menghitung BMI nya dari indikator weight dan length yang sudah ada.

Untuk missing value yang lainnya masih terhitung kecil dengan rata-rata missing value nya kurang dari 10, sehingga akan diatasi dengan mengisinya menggunakan median dan modus dari kolom yang memiliki missing value tersebut.

In [6]:
# mengisi kolom BMI dengan menggunakan kolom weight dan length yang sudah ada
df['BMI'] = df['Weight'] / (df['Length'] / 100) ** 2

In [7]:
# Mengatasi missing values
df.fillna(df.median(numeric_only=True), inplace=True)  # Mengisi dengan median untuk kolom numerik
df.fillna(df.mode().iloc[0], inplace=True)  # Mengisi dengan modus untuk kolom kategorikal

Untuk missing value dengan data numerik akan diatasi dengan median dari kalom tersebut agar nilai nya tidak bias (menjadi outlier).

Untuk missing value dengan data kategorikal akan diatasi dengan modus dari kolom tersebut agar data yang kosong terisi dengan nilai yang paling umum di kolom tersebut.

In [8]:
df.isnull().sum()

No                       0
Age                      0
Weight                   0
Length                   0
Sex                      0
BMI                      0
DM                       0
HTN                      0
Current Smoker           0
EX-Smoker                0
FH                       0
Obesity                  0
CRF                      0
CVA                      0
Airway disease           0
Thyroid Disease          0
CHF                      0
DLP                      0
BP                       0
PR                       0
Edema                    0
Weak Peripheral Pulse    0
Lung rales               0
Systolic Murmur          0
Diastolic Murmur         0
Typical Chest Pain       0
Dyspnea                  0
Function Class           0
Atypical                 0
Nonanginal               0
Exertional CP            0
LowTH Ang                0
Q Wave                   0
St Elevation             0
St Depression            0
Tinversion               0
LVH                      0
Poor R Progression       0
FBS                      0
CR                       0
TG                       0
LDL                      0
HDL                      0
BUN                      0
ESR                      0
HB                       0
K                        0
Na                       0
WBC                      0
Lymph                    0
Neut                     0
PLT                      0
EF-TTE                   0
Region RWMA              0
VHD                      0
Cath                     0
dtype: int64

Terlihat bahwa dataset saat ini sudah tidak lagi memiliki missing value pada setiap indikator nya.

In [9]:
# menampilkan kembali dataset yang missing value nya telah teratasi
df.head(20)

No  Age  Weight  Length    Sex        BMI DM HTN Current Smoker EX-Smoker  \
0    1   53      90     175   Male  29.387755  n   Y              Y         n   
1    2   67      70     157  Fmale  28.398718  n   Y              n         n   
2    3   54      54     164   Male  20.077335  n   n              Y         n   
3    4   66      67     158  Fmale  26.838648  n   Y              n         n   
4    5   50      87     153  Fmale  37.165193  n   Y              n         n   
5    6   50      75     175   Male  24.489796  n   n              Y         n   
6    7   55      80     165   Male  29.384757  n   n              n         Y   
7    8   72      80     175   Male  26.122449  Y   n              Y         n   
8    9   58      84     163  Fmale  31.615793  n   n              n         n   
9   10   60      71     170   Male  24.567474  Y   n              n         n   
10  11   58      75     168   Male  26.573129  n   Y              n         Y   
11  12   80      67     153  Fmale  28.621470  n   Y              n         n   
12  13   70      70     151  Fmale  30.700408  Y   Y              n         n   
13  14   67      63     154  Fmale  26.564345  Y   Y              n         n   
14  15   66      63     155  Fmale  26.222685  Y   Y              n         n   
15  16   59      81     167   Male  29.043709  Y   n              n         n   
16  17   41      68     169   Male  23.808690  n   n              Y         n   
17  18   68      59     161  Fmale  22.761468  n   n              n         n   
18  19   60      89     163  Fmale  33.497685  Y   Y              n         n   
19  20   65      72     150  Fmale  32.000000  Y   Y              n         n   

   FH Obesity CRF CVA Airway disease Thyroid Disease CHF DLP     BP   PR  \
0   n       Y   N   N              N               N   N   Y  110.0   80   
1   n       Y   N   N              N               N   N   N  140.0   80   
2   n       N   N   N              N               N   N   N  100.0  100   
3   n       Y   N   N              N               N   N   N  100.0   80   
4   n       Y   N   N              N               N   N   N  110.0   80   
5   n       N   N   N              N               N   N   N  118.0   70   
6   n       Y   N   N              N               N   N   N  110.0   80   
7   n       Y   N   N              N               N   N   Y  130.0   70   
8   n       Y   N   N              N               N   N   N   90.0   50   
9   n       N   N   N              N               N   N   N  130.0   70   
10  n       Y   N   N              N               N   N   N  170.0   70   
11  n       Y   N   N              N               N   N   Y  140.0  100   
12  n       Y   N   N              N               N   N   Y  140.0   84   
13  n       Y   N   N              N               N   N   Y  150.0   74   
14  n       Y   N   N              N               N   N   Y  130.0   80   
15  n       Y   N   N              N               N   N   Y  120.0   70   
16  n       N   N   N              N               N   N   N  130.0   80   
17  Y       N   N   N              N               N   N   Y  115.0   70   
18  n       Y   N   N              N               N   N   N  110.0   74   
19  n       Y   Y   N              N               N   N   Y  150.0   70   

   Edema Weak Peripheral Pulse Lung rales Systolic Murmur Diastolic Murmur  \
0     NO                     N          N               N                N   
1    Yes                     N          N               N                N   
2     NO                     N          N               N                N   
3     NO                     N          N               N                Y   
4     NO                     N          N               Y                N   
5     NO                     N          N               N                N   
6     NO                     N          N               Y                N   
7     NO                     N          N               N                N   
8  

**MENGUBAH SEMUA KOLOM MENJADI BERNILAI NUMERIK**

In [10]:
from sklearn.preprocessing import LabelEncoder, MinMaxScaler

In [11]:
# Mengubah semua kolom menjadi bernilai numerik
label_encoders = {}
for col in df.select_dtypes(include=['object']).columns:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col].astype(str))
    label_encoders[col] = le

Semua kolom yang ada pada dataset telah diubah menjadi bernilai numerik semua agar bisa dilakukan normalisasi. Karena jika ada kolom yang berisi teks, maka proses normalisasi tidak bisa dilakukan/error.

In [12]:
# menampilkan kembali data yang sudah diubah menjadi nilai numerik semua
display(df.head(20))

No  Age  Weight  Length  Sex        BMI  DM  HTN  Current Smoker  \
0    1   53      90     175    1  29.387755   1    0               0   
1    2   67      70     157    0  28.398718   1    0               2   
2    3   54      54     164    1  20.077335   1    1               0   
3    4   66      67     158    0  26.838648   1    0               2   
4    5   50      87     153    0  37.165193   1    0               2   
5    6   50      75     175    1  24.489796   1    1               0   
6    7   55      80     165    1  29.384757   1    1               2   
7    8   72      80     175    1  26.122449   0    1               0   
8    9   58      84     163    0  31.615793   1    1               2   
9   10   60      71     170    1  24.567474   0    1               2   
10  11   58      75     168    1  26.573129   1    0               2   
11  12   80      67     153    0  28.621470   1    0               2   
12  13   70      70     151    0  30.700408   0    0               2   
13  14   67      63     154    0  26.564345   0    0               2   
14  15   66      63     155    0  26.222685   0    0               2   
15  16   59      81     167    1  29.043709   0    1               2   
16  17   41      68     169    1  23.808690   1    1               0   
17  18   68      59     161    0  22.761468   1    1               2   
18  19   60      89     163    0  33.497685   0    0               2   
19  20   65      72     150    0  32.000000   0    0               2   

    EX-Smoker  FH  Obesity  CRF  CVA  Airway disease  Thyroid Disease  CHF  \
0           1   1        1    0    0               0                0    0   
1           1   1        1    0    0               0                0    0   
2           1   1        0    0    0               0                0    0   
3           1   1        1    0    0               0                0    0   
4           1   1        1    0    0               0                0    0   
5           1   1        0    0    0               0                0    0   
6           0   1        1    0    0               0                0    0   
7           1   1        1    0    0               0                0    0   
8           1   1        1    0    0               0                0    0   
9           1   1        0    0    0               0                0    0   
10          0   1        1    0    0               0                0    0   
11          1   1        1    0    0               0                0    0   
12          1   1        1    0    0               0                0    0   
13          1   1        1    0    0               0                0    0   
14          1   1        1    0    0               0                0    0   
15          1   1        1    0    0               0                0    0   
16          1   1        0    0    0               0                0    0   
17          1   0        0    0    0               0                0    0   
18          1   1        1    0    0               0                0    0   
19          1   1        1    1    0               0                0    0   

    DLP     BP   PR  Edema  Weak Peripheral Pulse  Lung rales  \
0     1  110.0   80      0                      0           0   
1     0  140.0   80      1                      0           0   
2     0  100.0  100      0                      0           0   
3     0  100.0   80      0                      0           0   
4     0  110.0   80      0                      0           0   
5     0  118.0   70      0                      0           0   
6     0  110.0   80      0                      0           0   
7     1  130.0   70      0                      0           0   
8     0   90.0   50      0                      0           0   
9     0  130.0   70      0                      0           0   
10    0  170.0   70      0                      0           0   
11    1  140.0  100      0                      0           0   
12    1

**MELAKUKAN NORMALISASI PADA MASING-MASING ATRIBUT**

Untuk normalisasi saya menggunakan MInMaxScaller, karena hanya akan mengubah skala data tanpa mengubah bentuk distribusinya dan rentang nilai yang lebih mudah di interpretasikan (0-1).

In [13]:
# Normalisasi data
scaler = MinMaxScaler()
df_scaled = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

In [14]:
# Menampilkan hasil normalisasi
display(df_scaled.head(20))

No       Age    Weight    Length  Sex       BMI   DM  HTN  \
0   0.000000  0.410714  0.583333  0.729167  1.0  0.494721  1.0  0.0   
1   0.003311  0.660714  0.305556  0.354167  0.0  0.451314  1.0  0.0   
2   0.006623  0.428571  0.083333  0.500000  1.0  0.086105  1.0  1.0   
3   0.009934  0.642857  0.263889  0.375000  0.0  0.382846  1.0  0.0   
4   0.013245  0.357143  0.541667  0.270833  0.0  0.836058  1.0  0.0   
5   0.016556  0.357143  0.375000  0.729167  1.0  0.279759  1.0  1.0   
6   0.019868  0.446429  0.444444  0.520833  1.0  0.494590  1.0  1.0   
7   0.023179  0.750000  0.444444  0.729167  1.0  0.351413  0.0  1.0   
8   0.026490  0.500000  0.500000  0.479167  0.0  0.592505  1.0  1.0   
9   0.029801  0.535714  0.319444  0.625000  1.0  0.283168  0.0  1.0   
10  0.033113  0.500000  0.375000  0.583333  1.0  0.371193  1.0  0.0   
11  0.036424  0.892857  0.263889  0.270833  0.0  0.461090  1.0  0.0   
12  0.039735  0.714286  0.305556  0.229167  0.0  0.552331  0.0  0.0   
13  0.043046  0.660714  0.208333  0.291667  0.0  0.370807  0.0  0.0   
14  0.046358  0.642857  0.208333  0.312500  0.0  0.355812  0.0  0.0   
15  0.049669  0.517857  0.458333  0.562500  1.0  0.479622  0.0  1.0   
16  0.052980  0.196429  0.277778  0.604167  1.0  0.249867  1.0  1.0   
17  0.056291  0.678571  0.152778  0.437500  0.0  0.203906  1.0  1.0   
18  0.059603  0.535714  0.569444  0.479167  0.0  0.675098  0.0  0.0   
19  0.062914  0.625000  0.333333  0.208333  0.0  0.609367  0.0  0.0   

    Current Smoker  EX-Smoker   FH  Obesity  CRF  CVA  Airway disease  \
0              0.0        0.5  1.0      1.0  0.0  0.0             0.0   
1              1.0        0.5  1.0      1.0  0.0  0.0             0.0   
2              0.0        0.5  1.0      0.0  0.0  0.0             0.0   
3              1.0        0.5  1.0      1.0  0.0  0.0             0.0   
4              1.0        0.5  1.0      1.0  0.0  0.0             0.0   
5              0.0        0.5  1.0      0.0  0.0  0.0             0.0   
6              1.0        0.0  1.0      1.0  0.0  0.0             0.0   
7              0.0        0.5  1.0      1.0  0.0  0.0             0.0   
8              1.0        0.5  1.0      1.0  0.0  0.0             0.0   
9              1.0        0.5  1.0      0.0  0.0  0.0             0.0   
10             1.0        0.0  1.0      1.0  0.0  0.0             0.0   
11             1.0        0.5  1.0      1.0  0.0  0.0             0.0   
12             1.0        0.5  1.0      1.0  0.0  0.0             0.0   
13             1.0        0.5  1.0      1.0  0.0  0.0             0.0   
14             1.0        0.5  1.0      1.0  0.0  0.0             0.0   
15             1.0        0.5  1.0      1.0  0.0  0.0             0.0   
16             0.0        0.5  1.0      0.0  0.0  0.0             0.0   
17             1.0        0.5  0.0      0.0  0.0  0.0             0.0   
18             1.0        0.5  1.0      1.0  0.0  0.0             0.0   
19             1.0        0.5  1.0      1.0  1.0  0.0             0.0   

    Thyroid Disease  CHF  DLP    BP        PR  Edema  Weak Peripheral Pulse  \
0               0.0  0.0  1.0  0.20  0.500000    0.0                    0.0   
1               0.0  0.0  0.0  0.50  0.500000    1.0                    0.0   
2               0.0  0.0  0.0  0.10  0.833333    0.0                    0.0   
3               0.0  0.0  0.0  0.10  0.500000    0.0                    0.0   
4               0.0  0.0  0.0  0.20  0.500000    0.0                    0.0   
5               0.0  0.0  0.0  0.28  0.333333    0.0                    0.0   
6               0.0  0.0  0.0  0.20  0.500000    0.0                    0.0   
7               0.0  0.0  1.0  0.40  0.333333    0.0                    0.0   
8               0.0  0.0  0.0  0.00  0.000000    0.0                    0.0   
9               0.0  0.0  0.0  0.40  0.333333    0.0                    0.0   
10              0.0  0.0  0.0  0.80  0.333333    0.0                    0.0   
11              0.0  0.0  1.0  0.50 

Dataset tersebut sudah dilakukan normalisasi menggunakan MinMaxScaller, sehingga semua nilai berada dalam rentang 0 hingga 1, tanpa mengubah distribusi data aslinya.